pybamm/solvers/dae_solver.py

```diff
@@ -73,7 +73,13 @@ def residuals(t, y, ydot):
             # turn into 1D arrays
             rhs_eval = rhs_eval[:, 0]
             alg_eval = alg_eval[:, 0]
-            return np.concatenate((rhs_eval - ydot[: rhs_eval.shape[0]], alg_eval))
+            # get entries of the mass matrix corresponding to rhs
+            mass_matrix = model.mass_matrix.entries[
+                0 : rhs_eval.shape[0], 0 : rhs_eval.shape[0]
+            ]
+            return np.concatenate(
+                (rhs_eval - mass_matrix @ ydot[: rhs_eval.shape[0]], alg_eval)
+            )
```

Comment:
`mass_matrix @ ydot` should have the same functionality (assuming `mass_matrix` has all zeros in the algebraic block).
Worth investigating which implementation is quicker, as this happens every time step

Reply:
yeah that is much faster thanks @tinosulzer !

Reply:
Np. I'd also put `model.mass_matrix.entries` outside of the `residuals` function